server/core/request.dylan

```diff
@@ -327,81 +356,41 @@ define inline function empty-line?
-  let parts = split(buffer, concatenate("--", boundary));
-  for (part in parts) 
-    let part = split(part, "\r\n\r\n");
-    let header-entries = split(first(part), "\r\n");
-    let disposition = #f;
-    let name = #f;
-    let type = #f;
-    let filename = #f;
-    for (header-entry in header-entries)
-      let header-entry-parts = split(header-entry, ';');
-      for (header-entry-part in header-entry-parts)
-        let eq-pos = char-position('=', header-entry-part, 0, size(header-entry-part));
-        let p-pos = char-position(':', header-entry-part, 0, size(header-entry-part));
-        if (p-pos & (substring(header-entry-part, 0, p-pos) = "Content-Disposition"))
-          disposition := substring(header-entry-part, p-pos + 2, size(header-entry-part));
-        elseif (p-pos & (substring(header-entry-part, 0, p-pos) = "Content-Type"))
-          type := substring(header-entry-part, p-pos + 2, size(header-entry-part));
-        elseif (eq-pos & (substring(header-entry-part, 0, eq-pos) = "name"))
-          // name unquoted
-          name := substring(header-entry-part, eq-pos + 2, size(header-entry-part) - 1);
-        elseif (eq-pos & (substring(header-entry-part, 0, eq-pos) = "filename"))
-          // filename unquoted
-          filename := substring(header-entry-part, eq-pos + 2, size(header-entry-part) - 1);
-        end if;
-      end for;
-    end for;
-    if (part.size > 1)
-      // TODO: handle disposition = "multipart/form-data" and parse that again
-      //disposition = "multipart/form-data" => ...
-      if (disposition = "form-data")
-        let content = substring(second(part), 0, size(second(part)) - 1);
-        request.request-query-values[name]
-          := if (filename & type)
-               make(<http-file>, filename: filename, content: content, mime-type: type);
-             else
-               content;
-             end if;
-      end if;
-    end if;
-  end for;
-end method extract-form-data;
-*/
-
 // Hey look!  More stuff to get rid of or move...
 
+// Query values are
+//   * URL parameters
+//   * multipart/form-data body content
+//   * application/x-www-form-urlencoded body content
+
+// get-query-value returns the first value for `key`. Use get-query-values if
+// you want a vector of all the values for that key. For example, form data for
+// multiple files uploaded under the same key.
 define inline function get-query-value
     (key :: <string>, #key as: as-type :: false-or(<type>))
  => (value :: <object>)
   let val = element(*request*.request-query-values, key, default: #f);
-  if (as-type & val)
+  let val = if (instance?(val, <list>) & ~empty?(val))
+              head(val)
+            else
+              val
+            end;
+  if (as-type & instance?(val, <string>))
     as(as-type, val)
   else
     val
   end
 end function get-query-value;
 
+define inline function get-query-values (key :: <string>) => (values :: <list>)
```

Comment:
Why not generic `<sequence>` for the API?

Reply:
Done, thanks.